### ISSUE TITLE
OS X builds

### ISSUE THREAD

**Issue by Palakis** [Palakis/obs-websocket#16]
Right now I can build the plugin on Windows and Linux. However, I need access to an OS X system to build and test an OS X version.

**Comment by asquelt** [comment]
osx build fails on mbedtls. i'd be grateful for any hints that may push this forward.
unfortunately i've got only macbook, so i can't provide you a builder host to use (but i can build manually on new releases).

[![asciicast](https://asciinema.org/a/3vz0sdegvsyq5dma33helyx3f.png)](https://asciinema.org/a/3vz0sdegvsyq5dma33helyx3f)

**Comment by Palakis** [comment]
@asquelt Seems like the problem lies here : https://github.com/Palakis/obs-websocket/blob/master/CMakeLists.txt#L46

**Comment by asquelt** [comment]
i did ```ln -s plugins/obs-websocket/deps/mbedtls deps/``` and i'm still getting the same error :(

**Comment by jnovack** [comment]
Does this help? http://www.wikigain.com/install-macos-sierra-10-12-vmware/

**Comment by Palakis** [comment]
@jnovack thanks for the link, but I already a macOS VM on my computer. Good to remind me of this issue though.

**Comment by jnovack** [comment]
Let me know if you need assistance somewhere with it some continuous integration or home-grown solutions.  A lot of mostly linux commands are available.. mostly.

**Comment by jnovack** [comment]
If you are really adventurous, Travis offers a OSX CI Environment.  Although I'm sure that getting a build set up script to install all your pre-reqs will take longer than you probably want it to. :)

https://docs.travis-ci.com/user/osx-ci-environment/

**Comment by RubyComputer** [comment]
@Palakis I can help you with testing in the upcoming days

**Comment by Palakis** [comment]
@asquelt I've just red your asciinema and noticed that you build obs-studio and obs-websocket together, with obs-websocket as a CMake submodule, Please try with a procedure similar to Windows and Linux, where obs-websocket is built as a third-party plugin.

**Comment by asquelt** [comment]
i've tried that first, without success, so i've switched to the method shown above.

**Comment by RubyComputer** [comment]
Still waiting on obs-websocket to be working on OSX. If you need any help just let me know

**Comment by asquelt** [comment]
@Palakis does this ring any bells?

https://travis-ci.org/asquelt/obs-websocket/jobs/218776363

do you want PR with Travis tests?

**Comment by Palakis** [comment]
@asquelt LIBOBS_LIB should point to a libobs.dylib file (or similar) that the compiler will link to. This is why you've got that error.

**Comment by Palakis** [comment]
@asquelt You need to build OBS Studio first because the compiler needs to link to libobs and the Frontend API. Once built, LIBOBS_LIB should point to a libobs.dylib file (or similar) that the compiler will link to. Or maybe remove it to see if OBS' CMake helpers find the library.

**Comment by asquelt** [comment]
i know, the link above shows log from entire build process, containing obs-studio. some of it might be collapsed, watch for triangle marks ;)

**Comment by Palakis** [comment]
@asquelt Oh, sorry. Gonna check this again.

**Comment by asquelt** [comment]
i've adjusted LIBOBS_LIB and it says
```-- Found Libobs: /Users/travis/build/asquelt/obs-studio/build/libobs  ```
but still throws linker exceptions.
last build: https://travis-ci.org/asquelt/obs-websocket/jobs/218833273

**Comment by Palakis** [comment]
@asquelt Look for a file named libobs.dylib in OBS' build dir or one of its subdirectories. And specify its path in LIBOBS_LIB.

**Comment by asquelt** [comment]
yep, that's already done.
see here: https://travis-ci.org/asquelt/obs-websocket/jobs/218833273#L2860
and here: https://travis-ci.org/asquelt/obs-websocket/jobs/218833273#L2875

**Comment by asquelt** [comment]
and this one is with auto-detection:
https://travis-ci.org/asquelt/obs-websocket/jobs/218842473
also bails out on linker.

**Comment by Palakis** [comment]
@asquelt Gonna try a build on my OS X VM to find the full path of libobs.dylib

**Comment by asquelt** [comment]
please do check, you should get two locations:
in build dir: obs-studio/build/libobs/libobs.dylib
after install in: /usr/local/bin/libobs.dylib

however cmake for obs-websocket get both directions right, so i'm not sure this is root cause.

**Comment by Palakis** [comment]
@asquelt Try with `-DLIBOBS_LIB=../../obs-studio/build/libobs/libobs.dylib`

**Comment by asquelt** [comment]
https://travis-ci.org/asquelt/obs-websocket/jobs/219214284

**Comment by asquelt** [comment]
there's linker invocation, maybe that will shed some light...
(line splitted for readability)
```
Apple LLVM version 8.1.0 (clang-802.0.38)
Target: x86_64-apple-darwin16.5.0
Thread model: posix
InstalledDir: /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin
 "/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/ld"
 -demangle
 -lto_library
 /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/lib/libLTO.dylib
 -dynamic
 -arch x86_64
 -bundle
 -macosx_version_min 10.12.0
 -o libobs-websocket.so
 -headerpad_max_install_names 
 CMakeFiles/obs-websocket.dir/obs-websocket.cpp.o
 CMakeFiles/obs-websocket.dir/WSServer.cpp.o 
 CMakeFiles/obs-websocket.dir/WSRequestHandler.cpp.o
 CMakeFiles/obs-websocket.dir/WSEvents.cpp.o
 CMakeFiles/obs-websocket.dir/Config.cpp.o
 CMakeFiles/obs-websocket.dir/Utils.cpp.o
 CMakeFiles/obs-websocket.dir/forms/settings-dialog.cpp.o
 CMakeFiles/obs-websocket.dir/obs-websocket_automoc.cpp.o
 -rpath
 /Users/travis/build/asquelt/obs-studio/build/libobs
 /Users/travis/build/asquelt/obs-studio/build/libobs/libobs.0.dylib
 /usr/local/opt/qt/lib/QtWebSockets.framework/QtWebSockets
 /usr/local/opt/qt/lib/QtWidgets.framework/QtWidgets
 deps/mbedtls/library/libmbedcrypto.a
 /usr/local/opt/qt/lib/QtNetwork.framework/QtNetwork
 /usr/local/opt/qt/lib/QtGui.framework/QtGui /usr/local/opt/qt/lib/QtCore.framework/QtCore
 -lc++ 
 -lSystem
 /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/../lib/clang/8.1.0/lib/darwin/libclang_rt.osx.a
```
entire log: https://travis-ci.org/asquelt/obs-websocket/jobs/221332667#L2966

**Comment by Palakis** [comment]
@asquelt I'm not an expert on OS X binaries but it's not normal that the build process is trying to build an .so on OS X. And it seems like the linker can't shared-link a binary to a .dylib

**Comment by Palakis** [comment]
@asquelt Is there an libobs.a in the same folder as libobs.dylib?

**Comment by asquelt** [comment]
there's no static `libobs.a` in project build. it seems that they only provide dynamic one.

**Comment by Palakis** [comment]
@asquelt Can you please submit a Pull Request of your work ? (both for Linux and OS X builds on Travis)

**Comment by Palakis** [comment]
I think I found the source of the linker errors. Because the OS X build instructions where not set in `CMakeLists.txt`, the link to OBS' frontend API wasn't specified and the linker failed. On the other hand, linking to the libobs symbols works fine with the .dylib of libobs.

**Comment by asquelt** [comment]
`ld: library not found for -llibobs-frontend-api` :(

**Comment by Palakis** [comment]
@asquelt Yup, working on this.

**Comment by Palakis** [comment]
It works! IT'S ALIVE! https://travis-ci.org/Palakis/obs-websocket/jobs/222239566

**Comment by ErikMinekus** [comment]
Great work! I was able to build it locally, but when trying to load it I got this error:
```
os_dlopen(../obs-plugins/obs-websocket.so->../obs-plugins/obs-websocket.so): dlopen(../obs-plugins/obs-websocket.so, 1): Library not loaded: /usr/local/opt/qt/lib/QtWebSockets.framework/Versions/5/QtWebSockets
  Referenced from: /Applications/OBS.app/Contents/Resources/obs-plugins/obs-websocket.so
  Reason: image not found

Module '../obs-plugins/obs-websocket.so' not found
```

I was able to get past this with `brew install qt5`, but I assume it means QtWebSockets was not properly linked?

Afterwards I got:
```
Failed to load 'en-US' text for module: 'obs-websocket.so'
[obs-websocket] you can haz websockets (version 4.0.0)
```

Followed by OBS crashing:
```
abort() called

Thread 0 Crashed:: Dispatch queue: com.apple.main-thread
0   libsystem_kernel.dylib        	0x00007fffd618bd42 __pthread_kill + 10
1   libsystem_pthread.dylib       	0x00007fffd62795bf pthread_kill + 90
2   libsystem_c.dylib             	0x00007fffd60f1420 abort + 129
3   org.qt-project.QtCore         	0x000000017774bd69 0x177736000 + 89449
4   org.qt-project.QtCore         	0x000000017774d6e0 QMessageLogger::fatal(char const*, ...) const + 224
5   org.qt-project.QtWidgets      	0x0000000176a5915c QWidgetPrivate::QWidgetPrivate(int) + 604
6   org.qt-project.QtWidgets      	0x0000000176c209de QDialog::QDialog(QWidget*, QFlags<Qt::WindowType>) + 46
7   obs-websocket.so              	0x00000001769c342b SettingsDialog::SettingsDialog(QWidget*) + 75
8   obs-websocket.so              	0x00000001769c46ad SettingsDialog::SettingsDialog(QWidget*) + 29
9   obs-websocket.so              	0x00000001769b9552 obs_module_load + 386
10  libobs.0.dylib                	0x0000000110c48f53 obs_init_module + 83
11  libobs.0.dylib                	0x0000000110c49e50 load_all_callback + 64
12  libobs.0.dylib                	0x0000000110c49d68 obs_find_modules + 2280
13  libobs.0.dylib                	0x0000000110c49463 obs_load_all_modules + 35
14  obs                           	0x000000010da1bec5 OBSBasic::OBSInit() + 757
15  obs                           	0x000000010da0a3a9 OBSApp::OBSInit() + 553
16  obs                           	0x000000010da0c737 main + 4711
17  libdyld.dylib                 	0x00007fffd605d235 start + 1
```

**Comment by Palakis** [comment]
> I was able to get past this with `brew install qt5`, but I assume it means QtWebSockets was not properly linked?

This means that Qt5 wasn't properly installed or installed at all.

Regarding this error :
```
Failed to load 'en-US' text for module: 'obs-websocket.so'
```
This isn't a fatal one: OBS tried to load the translation files but can't find them, so it won't load the language strings for the current locale and put whatever string it can put.

As for the crash, I'm gonna investigate this.

**Comment by ErikMinekus** [comment]
I installed Qt5 on a different Mac and built obs-studio and obs-websocket there, but it wasn't installed on my Mac with OBS Studio. You don't need to have Qt5 installed to run OBS Studio, so I would think it's also not needed for this plugin.

**Comment by Palakis** [comment]
@ErikMinekus Ah sorry, what this means is that Qt is embed with OBS Studio (like the Windows version) with only its required Qt modules. QtWebSockets then has to be added on the system or the OBS install for obs-websocket to work.

The OS X builds are a WIP but the binary releases will obviously have QtWebSockets embedded with them.

**Comment by Palakis** [comment]
@ErikMinekus BTW can you try to build and run obs-websocket on your system with the latest commits, please? (committed several minutes ago)

**Comment by ErikMinekus** [comment]
New stack trace:
```
0   libsystem_kernel.dylib        	0x00007fffd618bd42 __pthread_kill + 10
1   libsystem_pthread.dylib       	0x00007fffd62795bf pthread_kill + 90
2   libsystem_c.dylib             	0x00007fffd60f1420 abort + 129
3   org.qt-project.QtCore         	0x0000000170c29d69 0x170c14000 + 89449
4   org.qt-project.QtCore         	0x0000000170c2b6e0 QMessageLogger::fatal(char const*, ...) const + 224
5   org.qt-project.QtWidgets      	0x000000017008915c QWidgetPrivate::QWidgetPrivate(int) + 604
6   org.qt-project.QtWidgets      	0x00000001702509de QDialog::QDialog(QWidget*, QFlags<Qt::WindowType>) + 46
7   obs-websocket.so              	0x000000016fff6e98 SettingsDialog::SettingsDialog(QWidget*) + 72   <-- This was 75 in the last crash
8   obs-websocket.so              	0x000000016fff80fd SettingsDialog::SettingsDialog(QWidget*) + 29
9   obs-websocket.so              	0x000000016ffecf22 obs_module_load + 386
10  libobs.0.dylib                	0x000000010a177f53 obs_init_module + 83
11  libobs.0.dylib                	0x000000010a178e50 load_all_callback + 64
12  libobs.0.dylib                	0x000000010a178d68 obs_find_modules + 2280
13  libobs.0.dylib                	0x000000010a178463 obs_load_all_modules + 35
14  obs                           	0x0000000106f47ec5 OBSBasic::OBSInit() + 757
15  obs                           	0x0000000106f363a9 OBSApp::OBSInit() + 553
16  obs                           	0x0000000106f38737 main + 4711
17  libdyld.dylib                 	0x00007fffd605d235 start + 1
```

**Comment by Palakis** [comment]
OS X builds are now working quite good on Travis, along with packaging and upload to S3 : https://s3.eu-central-1.amazonaws.com/obs-websocket-osx-builds/obs-websocket-7bb8e56-master-osx.pkg

Still that startup crash to fix, but now that I have an OS X dev env setup and packages for anyone to test, it's gonna be much easier.

**Comment by ErikMinekus** [comment]
FYI if I comment out the code for SettingsDialog in obs-websocket.cpp, the plugin runs fine and I can connect to localhost:4444. I'm using macOS Sierra (10.12.4) and Qt 5.8 to build and run the plugin.

**Comment by Palakis** [comment]
@ErikMinekus The stack traces obviously show a problem with the `SettingsDialog`. It is either caused by a Qt version mismatch or a platform-specific behavior of Qt.

**Comment by asquelt** [comment]
after installing pkg from s3 i'm getting this:
```error: os_dlopen(/Library/Application Support/obs-studio/plugins/obs-websocket/bin/obs-websocket.so->/Library/Application Support/obs-studio/plugins/obs-websocket/bin/obs-websocket.so): dlopen(/Library/Application Support/obs-studio/plugins/obs-websocket/bin/obs-websocket.so, 1): Library not loaded: /usr/local/opt/qt/lib/QtWebSockets.framework/Versions/5/QtWebSockets
  Referenced from: /Library/Application Support/obs-studio/plugins/obs-websocket/bin/obs-websocket.so
  Reason: image not found

warning: Module '/Library/Application Support/obs-studio/plugins/obs-websocket/bin/obs-websocket.so' not found
```

**Comment by Palakis** [comment]
@asquelt Yes I'm aware of this. I need to figure out how to include QtWebSockets' binary from the Qt5 path into the package.

And figure out why obs-websocket crashes because of the settings window.

**Comment by Palakis** [comment]
@ErikMinekus The obs-websocket OS X binary is in fact not linked correctly. Sorry for my answer, I didn't know at the time that absolute paths for linked shared libraries is a thing on OS X...

I managed to include the QtWebSockets library into the package but I'm currently struggling with having obs-websocket link to an rpath instead of an absolute path.

**Comment by jnovack** [comment]
```bash
$ install_name_tool
Usage: /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/install_name_tool [-change old new] ... [-rpath old new] ... [-add_rpath new] ... [-delete_rpath old] ... [-id name] input
```

Or for short.
`install_name_tool -change <from> <to> <executable>`

You can do this within the linker.

One concrete option that works would be to set the install_name flag when linking the .dylib.

`gcc -dynamiclib -install_name '$(CURDIR)/hidelib/libmylibrary.dylib' -current_version 1.0 mylibrary.o -o libmylibrary.dylib`

Then you can just link to the library normally:

`gcc main.o -L '$(CURDIR)/hidelib' -lmylibrary -o main`

And finally, a VERY long article on rpath linking: https://github.com/conda/conda-build/issues/279

References:

* http://stackoverflow.com/questions/10021428/macos-how-to-link-a-dynamic-library-with-a-relative-path-using-gcc-ld

**Comment by Palakis** [comment]
@jnovack That's what I did before you posted this. This also is the way OBS Studio its OS X package.

The OS X version doesn't crash anymore on startup. But the server doesn't start properly and the locale files are not found.

**Comment by Palakis** [comment]
Server issue fixed!  
Remains to fix : locale issues and another problem with linking.

**Comment by Palakis** [comment]
Linking problem fixed, still that locale issue to fix. Here's a working build : https://obs-websocket-osx-builds.s3-eu-central-1.amazonaws.com/obs-websocket-5fbc2cb-master-osx.pkg

**Comment by Palakis** [comment]
Locale issue likely fixed! Here's the current latest build : https://obs-websocket-osx-builds.s3-eu-central-1.amazonaws.com/obs-websocket-bc24497-master-osx.pkg

**Comment by Palakis** [comment]
@asquelt @jnovack @ErikMinekus @RubyComputer Can you please do some testing on your end with the binary build for OS X mentioned above?

**Comment by asquelt** [comment]
works fine without password (i've tried t2t2).
it throws exception with password enabled and set: https://gist.github.com/asquelt/6af1b25969a90e7efd80d426115aebff

**Comment by Palakis** [comment]
@asquelt Exception thrown when setting the password, when authenticating or when sending a new request?

**Comment by asquelt** [comment]
after n'th request from client. sometimes t2t2 renders the page and loads up the scenes, but eventually it blows. sometimes it blows immediately after providing password to t2t2.

**Comment by ErikMinekus** [comment]
I did not set a password and I'm getting the same random crashes by just clicking around in t2t2:
https://gist.github.com/ErikMinekus/f4ab7837dc96a10a8e72d5d565481440

**Comment by Palakis** [comment]
All stack traces from @asquelt and @ErikMinekus have in common that `incorrect checksum for freed object - object was probably modified after being freed.` error...

**Comment by ErikMinekus** [comment]
This crash from build 82d74fc seems to more clearly crash in `WSRequestHandler::SendOKResponse()`, maybe that helps to pinpoint the issue: https://gist.github.com/ErikMinekus/899e6d9be0c101ef2d4df120d46f4842

**Comment by Palakis** [comment]
@ErikMinekus looks more like a bug in `tiny_free_list_remove_ptr`

**Comment by ErikMinekus** [comment]
In Config.cpp:
```
unsigned char *passAndSalt = (unsigned char*)bzalloc(passwordLength + saltLength);
unsigned char *challenge = (unsigned char*)bzalloc(64);
char *challengeAndResponse = (char*)bzalloc(secretLength + sessChallengeLength);
unsigned char *expected_response = (unsigned char*)bzalloc(64);
```
Shouldn't all these string sizes be `+ 1` to account for the null character? A couple of lines later you try to null-terminate the strings by writing outside their memory.

**Comment by jnovack** [comment]
Sorry, just got back to the conversation.

Agree with Erik. 

```cpp
	// Encode the SHA256 hash to Base64
	unsigned char *expected_response = (unsigned char*)bzalloc(64);
	size_t base64_size = 0;
	mbedtls_base64_encode(expected_response, 64, &base64_size, hash, 32);
	expected_response[64] = 0; // Null-terminate the string
```

`expected_response` is of `size()/length()` 65 if you use `expected_response[64]`.  0 is a number, dammit! 

Additionally, and this is for consistency and mostly preference; while all nulls are equal, there are double-plus ungood edge cases when some nulls are less equal than others.  I humbly suggest:

```diff
-expected_response[64] = 0; // Null-terminate the string
+expected_response[64] = '\0'; // Null-terminate the string
```

**Comment by Palakis** [comment]
>  We're only dealing with unsigned chars, not strings. Unless I'm SERIOUSLY missing something.

Null-terminating is done only on arrays of `char` representing "strings". These "strings" are of type `unsigned char*` because mbedtls_* function use this type both for "strings" and binary data, and therefore to save a cast. 
But I should clearly differentiate text as `char*` (or `const char*`) and binary as `unsigned char*`, even though these two types are supposed to be the same on a modern Intel/AMD CPU.

---

Regarding the arithmetic of strings here... This is something that I didn't took time to re-check, and as I struggled with mbedtls to have something working for authentication at the first place. This is bad, and I'll take time to fix this. Please acknowledge the massive amount of work I do on this project, so mistakes can happen (even though I prefer no mistakes at all 😄 ).

**Comment by Palakis** [comment]
@ErikMinekus Regarding the out-of-memory write, this is also something I noticed, but I maybe was too tired to wonder why this worked without segfaulting... :thinking:

**Comment by Palakis** [comment]
Turns out mbedtls outputs a correct string with a null termination when calling `mbedtls_base64_encode`, so no need to null-terminate it manually...

**Comment by Palakis** [comment]
Issue addressed in https://github.com/Palakis/obs-websocket/commit/fba9bd2b763487ea5949bec1bbf665d939f08b7d. Thanks a lot for pointing this out! 😄

Back to those OS X crashes: It seems like a general problem within the plugin or OBS at the binary level, so maybe there's a version mismatch again (brew has Qt 5.8, both obs-websocket and OBS Studio are supposed to have that version, except if OBS Studio OS X packages are not from Travis...)

**Comment by jnovack** [comment]
> Please acknowledge the massive amount of work I do on this project, so mistakes can happen (even though I prefer no mistakes at all 😄 ).

I, in no way shape or form, wanted to project negative comments, my sincere apologies if you felt I was. You are doing a service to the OBS community (and to the world, by way of stream viewers :)) and I merely wanted to critically discuss the issue, in terms of "deep analysis", not "finding fault".

**Comment by ErikMinekus** [comment]
After making these changes to commit fba9bd2: https://gist.github.com/ErikMinekus/3e89bda57aaef84d427bc532b718813c

I couldn't get OBS to crash by clicking around in t2t2 for 10 minutes :smile: So it seems releasing data in the destructor is not a good idea either.

P.S. I'm definitely not well-versed in C++, so I'm just pointing out things that I think may be wrong :wink: Of course your work is appreciated 👍

**Comment by Palakis** [comment]
@jnovack Your feedback is appreciated! 😀 Apologies from my end if my comment seemed a bit harsh.

@ErikMinekus Maybe the bad code in Config.cpp caused some form of heap or stack corruption, or some other low-level bug...

**Comment by Palakis** [comment]
@ErikMinekus Sorry, didn't see those are changes on your end. Can you please test the latest OS X build please ? https://obs-websocket-osx-builds.s3-eu-central-1.amazonaws.com/obs-websocket-fba9bd2-master-osx.pkg

I see that you commented the call to `obs_data_release` when the request handler gets destroyed. If crashes still happen with the build mentioned above, this may mean that this call could be responsible for the OS X crashes...

**Comment by ErikMinekus** [comment]
fba9bd2 was still crashing, so I increased those string sizes (not sure if it also has to be done in the mbedtls_* calls), after which it started crashing in WSRequestHandler's destructor.

Just tried the latest build with and without a password and it does not crash.

**Comment by Palakis** [comment]
@ErikMinekus I removed the call to `obs_data_release` happening in `WSRequestHandler`'s destructor. It wasn't very logical to have this call here, as it is already released at the end of `processIncomingMessage` and shouldn't have a refcount higher than 1. Bad coding again...

**Comment by Palakis** [comment]
@ErikMinekus When I tested the "latest" build, it crashed in my El Capitan VM. But commenting `obs_data_release` make known crashes disappear.

In https://github.com/Palakis/obs-websocket/commit/e647debcfb372544e05f6556bac64c1f0500cf2c, I removed the call to `obs_data_release` happening in `WSRequestHandler`'s destructor. It wasn't very logical to have this call here, as it is already released at the end of `processIncomingMessage` and shouldn't have a refcount higher than 1. 

And so here's the build associated with this commit: https://obs-websocket-linux-builds.s3-eu-central-1.amazonaws.com/obs-websocket_1-e647deb-master-git-1_amd64.deb

**Comment by asquelt** [comment]
https://obs-websocket-osx-builds.s3-eu-central-1.amazonaws.com/obs-websocket-e647deb-master-osx.pkg for mac.

**Comment by Palakis** [comment]
@asquelt Woops. Fixed.

**Comment by asquelt** [comment]
good job Stephane! it's rock solid now.

**Comment by Palakis** [comment]
Thanks to everyone who helped to make this possible!
